### PR TITLE
Use Local Type Inference approach to typecheck polymorphic calls

### PR DESCRIPTION
**Pull request by xxdavid**
Hi!

After months of working on this, I finally come up with a PR that implements the ideas captured in #553. It behaves in the same way as I described in https://github.com/josefs/Gradualizer/issues/553#issuecomment-1934551242, I only changed the second error message a tiny bit.

The main changes are in the commit b00cadd2. It implements the new approach to typechecking polymorphic calls based on [Local Type Inference](https://www.cis.upenn.edu/~bcpierce/papers/lti-toplas.pdf). I think it is quite straightforward but I already know the algorithm well so I am probably biased. However, it should be at least much simpler than the original constraint solver. I tried to document it in the [wiki](https://github.com/josefs/Gradualizer/wiki/Polymorphism) (together with the rigid type variables), so you may read it as documentation to the code. I'm also writing a thesis about this change, so I will link it to the wiki when it's out, to provide another source of documentation.

The main benefit of this change should be error messages that are understandable even when polymorphism is involved. Only in a few cases, a more complex error message shows up but even then, it should be clear what's going on. I think that if this new approach proves successful, we can even enable `--solve_constraints` by default in the future.

As a byproduct, all the "shape limitations" of the original constraint solver are gone (this could have been also solved by fixing the constraint generation). A type variable can now be instantiated to both a tuple and an atom without any problems. 

Other big changes are in a3029e1a. In this commit, I remove the constraints and clauses for flexible type variables from most of the code. Constraints are now generated and propagated only in functions related to typechecking polymorphic calls and functions related to subtyping. This should make the code simpler and easier to read. It also trivially solves all but one "don't drop the constraints" TODOs because constraints are now combined only in the subtyping functions and in `type_check_poly_call/4`. (Please do not squash commits when merging the PR as this commit would make the resulting commit unreadable.)

I also fixed some things along the way (fc432f2c, 2323f102, e9a0afa4) and added a lot of tests.

While the new approach is better in some cases (like the `minus/1` in `poly_fail.erl`), it is also weaker in some other cases. I see the biggest limitation in that the type of anonymous functions passed to polymorphic functions is now inferred via `type_check_expr/2` and not `type_check_expr_in/2` as before, and thus some errors may not be discovered. This can, however, be solved in the future by solving the constraints gathered from all other arguments first and then typechecking the anonymous function using the information from the solved constraints. But I think it is better to have understandable errors than just more of them. Once we have this, let's then focus on inferring the type of anonymous functions passed to polymorphic functions.

### CODE REVIEW

**Comment by xxdavid** [comment]
@zuiderkwast @erszcz I know you're busy, but do you think you'll have time for a code review?

**Comment by erszcz** [comment]
@xxdavid Sorry for the lack of responsiveness. Life got surprisingly busy the last few months. I can imagine it being quite frustrating given the terrific effort you've put into this work.

That being said, I've just gone through the wiki description - it's excellent!
I'll also go over the code of the PR, in the next 1-2 days, and try it out on some examples. Please bear with me.

**Comment by xxdavid** [comment]
No need to apologize! :) It's a volunteer project after all and I can imagine life getting busy.

Thank you for your kind words and for planning to go over it. I'll stay tuned.

**Review by erszcz — APPROVED** [review]
This PR is a fantastic piece of work, it's truly impressive!

I've left a few superficial comments, there's nothing serious I could find.

I've also tried it out one some examples and the new and improved error messages are great!

File: src/gradualizer_fmt.erl [hunk 1]
```diff
@@ -350,14 +350,27 @@ format_type_error({unsupported_expression, Anno, Expr}, Opts) ->
...
+    FunExpr = {'fun', Anno, absform:extract_function_from_call(Expr)},
+    FormattedName = [pp_expr(Name, Opts), "/", integer_to_list(length(Args))],
+    ArgsWithTypes = lists:zip(Args, ArgTys),
+    Bindings = [{FunExpr, FunTy} | ArgsWithTypes],
+    FormattedBindings = lists:map(fun ({Arg, ArgTy}) ->
+        io_lib:format("    ~s :: ~s~n", [pp_expr(Arg, Opts), pp_type(ArgTy, Opts)])
+    end, Bindings),
+    io_lib:format(
+      "~sThe type variable ~s in the call to ~s~s is instantiated to both ~s and ~s,"
+      ++ " while the former should be a subtype of the latter, but it is not."
```
[the remark this answers is not in the document]
Comment (xxdavid) [reply src/gradualizer_fmt.erl]: Okay, I've split it into two sentences.

File: src/constraints.hrl [hunk 2]
```diff
@@ -1,10 +1,9 @@
...
 -record(constraints, { lower_bounds = #{} :: #{ constraints:var() =>
-                                                [gradualizer_type:abstract_type()] },
-                       upper_bounds = #{} :: #{ constraints:var() =>
-                                                [gradualizer_type:abstract_type()] },
-                       exist_vars   = #{} :: constraints:mapset(constraints:var()) }).
+                                                gradualizer_type:abstract_type() },
+                           upper_bounds = #{} :: #{ constraints:var() =>
+                                                gradualizer_type:abstract_type() } }).
 
 -endif. %% __CONSTRAINTS_HRL__
```
[the remark this answers is not in the document]
Comment (xxdavid) [reply src/constraints.hrl:9]: Thanks for catching! I've changed all `CONSTRAINTS_HRL__` to `__ CONSTRAINTS_HRL__`.

File: src/typechecker.erl [hunk 3]
```diff
@@ -123,11 +125,9 @@
 
 -include_lib("stdlib/include/assert.hrl").
 -include("constraints.hrl").
+-include("constraints.hrl").
```
[the remark this answers is not in the document]
Comment (xxdavid) [reply src/typechecker.erl]: Thanks for catching, removed.

File: src/typechecker.erl [hunk 4]
```diff
@@ -1746,6 +1755,90 @@ subst_ty(Sub, {ann_type, P, [AnnoVar, Type]}) ->
...
 
+-type variance() :: covariant | contravariant | invariant.
+-type variance_map() :: #{atom() => variance()}.
+
+%% Gathers all flexible type variables in a given type together with their variance
+%% with respect to the given type.
+%%
+%% Suppose R is a type that contains a type variable X.
+%% R is covariant in X if for every S and T such that S <: T, it holds that [S/X]R <: [T/X]R.
+%% R is contravariant in X if for every S and T such that S <: T, it holds that [T/X]R <: [R/X]R.
```
[the remark this answers is not in the document]
Comment (xxdavid) [reply src/typechecker.erl]: Of course, you're right! Fixed.

File: src/typechecker.erl [hunk 5]
```diff
@@ -640,61 +692,46 @@ glb(T1, T2, A, Env) ->
...
 %% Union types: glb distributes over unions
 glb_ty({type, Ann, union, Ty1s}, Ty2, A, Env) ->
-    {Tys, Css} = lists:unzip([ glb(Ty1, Ty2, A, Env) || Ty1 <- Ty1s ]),
-    {{type, Ann, union, Tys}, constraints:combine(Css)};
+    Tys = [ glb(Ty1, Ty2, A, Env) || Ty1 <- Ty1s ],
+    {type, Ann, union, Tys};
 glb_ty(Ty1, {type, Ann, union, Ty2s}, A, Env) ->
-    {Tys, Css} = lists:unzip([glb(Ty1, Ty2, A, Env) || Ty2 <- Ty2s ]),
-    {{type, Ann, union, Tys}, constraints:combine(Css)};
+    Tys = [glb(Ty1, Ty2, A, Env) || Ty2 <- Ty2s ],
```
[the remark this answers is not in the document]
Comment (xxdavid) [reply src/typechecker.erl]: A space added.

File: src/typechecker.erl [hunk 6]
```diff
@@ -2293,85 +2382,140 @@ type_check_rel_op(Env, Op, P, Arg1, Arg2) ->
...
+    %% namely the Local Type Argument Synthesis algorithm from the Local Type Inference paper
+    %% by Pierce and Turner (https://www.cis.upenn.edu/~bcpierce/papers/lti-toplas.pdf).
+    %% In essence it goes like this:
+    %% 1. check that all argument types are subtypes of the expected parameter types
+    %%    and gather constraints (upper and lower bounds for each type variables)
+    %%    along the way
+    %% 2. combine the bounds (lower bounds with LUB, upper bounds with GLB)
+    %% 3. check that the constraints are satisfiable (i.e., all lower bounds are subtypes
+    %%    of corresponding upper bounds), and if not, throw a type error
+    %% 4. compute a substitution that minimizes the result type (and thus preverves
```
[the remark this answers is not in the document]
Comment (xxdavid) [reply src/typechecker.erl]: Fixed, thanks!

**Comment by xxdavid** [comment]
Thank you very much for the review! :) I've fixed all the issues and resolved the threads.

Can I merge this pull request now? Or would you also like to take a look, @zuiderkwast?

File: src/typechecker.erl [hunk 7]
```diff
@@ -184,10 +182,23 @@ compatible(Ty1, Ty2, Env) ->
...
--spec subtype(type(), type(), env()) -> compatible().
+-spec subtype(type(), type(), env()) -> boolean().
 subtype(Ty1, Ty2, Env) ->
+    case subtype_with_constraints(Ty1, Ty2, Env) of
+        {true, _Cs} ->
+            %% There shouldn't be any constraint really.
+            %% We could assert that the constraints are empty but this could crash
+            %% Gradualizer even for end users if there was some error in our code,
+            %% it's probably safer not to assert and just drop the constraints
+            true;
```
Comment (zuiderkwast) [inline src/typechecker.erl:193]: Debug logging in verbose mode, would that be useful?

**Review by zuiderkwast — APPROVED** [review]
I don't have time to review properly, but I agree it's impressive.

**Comment by erszcz** [comment]
I think it's good to merge now 👍

BTW, @xxdavid, it was great to see you in person on Lambda Days!

File: src/typechecker.erl (hunk 7 above)
Comment (xxdavid) [reply src/typechecker.erl:193]: Hmm, good idea! I'll try that when I get to some other work on Gradualizer.

**Comment by xxdavid** [comment]
@erszcz Thanks for merging the PR, and it was really nice to see you, too!

**Pull request merged**